File: erp-finance-system/backend/finance-core-service/src/main/java/com/erp/finance/entity/APInvoice.java
package com.erp.finance.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "ap_invoices")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class APInvoice {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, unique = true, length = 100)
    private String invoiceNumber;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "vendor_id", nullable = false)
    private Vendor vendor;

    @Column(nullable = false, precision = 15, scale = 2)
    private BigDecimal amount;

    @Column(precision = 15, scale = 2)
    private BigDecimal taxAmount;

    @Column(precision = 15, scale = 2)
    private BigDecimal totalAmount;

    @Column(nullable = false)
    private LocalDate invoiceDate;

    @Column(nullable = false)
    private LocalDate dueDate;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private APInvoiceStatus status = APInvoiceStatus.PENDING;

    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private ApprovalStatus approvalStatus = ApprovalStatus.PENDING;

    @OneToMany(mappedBy = "invoice", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<APInvoiceLine> lines = new ArrayList<>();

    @Column(name = "approved_by")
    private UUID approvedBy;

    @Column(name = "approved_at")
    private LocalDateTime approvedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    public boolean isOverdue() {
        return status == APInvoiceStatus.PENDING && LocalDate.now().isAfter(dueDate);
    }
}

File: erp-finance-system/mobile/erp_finance_mobile/lib/models/ap_invoice.dart
class APInvoice {
  final String id;
  final String invoiceNumber;
  final String vendorName;
  final double amount;
  final double totalAmount;
  final String status;
  final String approvalStatus;
  final DateTime? dueDate;
  final String? description;

  APInvoice({
    required this.id,
    required this.invoiceNumber,
    required this.vendorName,
    required this.amount,
    required this.totalAmount,
    required this.status,
    required this.approvalStatus,
    this.dueDate,
    this.description,
  });

  factory APInvoice.fromJson(Map<String, dynamic> json) {
    return APInvoice(
      id: json['id']?.toString() ?? '',
      invoiceNumber: (json['invoiceNumber'] ?? json['invoice_number']) as String,
      vendorName: (json['vendorName'] ?? json['vendor_name']) as String,
      amount: (json['amount'] as num).toDouble(),
      totalAmount: json['totalAmount'] != null
          ? (json['totalAmount'] as num).toDouble()
          : (json['total_amount'] as num?)?.toDouble() ?? (json['amount'] as num).toDouble(),
      status: json['status']?.toString() ?? '',
      approvalStatus: (json['approvalStatus'] ?? json['approval_status'])?.toString() ?? '',
      dueDate: json['dueDate'] != null
          ? DateTime.parse(json['dueDate'] as String)
          : json['due_date'] != null
              ? DateTime.parse(json['due_date'] as String)
              : null,
      description: json['description'] as String?,
    );
  }
}
